common/error.h

```diff
@@ -71,19 +73,76 @@ class [[nodiscard]] Error : public Printable<Error> {
   std::string message_;
 };
 
-// Holds a value of type `T`, or an Error explaining why the value is
+// A common base class that custom error types should derive from.
+//
+// This combines the ability to be printed with the ability to convert the error
+// to a string and in turn to a non-customized `Error` type by rendering into a
+// string.
+//
+// The goal is that custom error types can be used for errors that are common
+// and/or would have cost to fully render the error message to a string. A
+// custom type can then be used to allow custom, light-weight handling of errors
+// when appropriate. But to avoid these custom types being excessively viral, we
+// ensure they can be converted to normal `Error` types when needed by rendering
+// fully to a string.
+template <typename ErrorT>
+class [[nodiscard]] ErrorBase : public Printable<ErrorT> {
+ public:
+  // NOLINTNEXTLINE(bugprone-crtp-constructor-accessibility): Deleted.
+  ErrorBase(const ErrorBase&) = delete;
+
+  auto ToString() const -> std::string {
+    RawStringOstream s;
+    static_cast<const ErrorT*>(this)->Print(s);
+    return s.TakeStr();
```

Comment:
I think `Printable::Dump` is identical, maybe you can use it directly? (assuming you want to keep the `ToString` name)

```suggestion
    return Dump();
```

Reply:
The `Dump` method is private, and I wasn't sure we wanted to add a dependency on it... No strong opinion about how to do this though -- the above just didn't compile.